Modernize python

rhaschke opened the pull request:
- Use find_package(Python)
- Use `pip install` instead of deprecated `python setup.py install`
  Drawback: pip performs an in-place build, creating an `.egg-info` folder. This is deleted after installation.

rhaschke committed:
Update find_package(PythonInterp)

rhaschke committed:
Bump minimum cmake version to 3.16

rhaschke committed:
Fix deprecated install procedure

Use "pip install" instead of "python setup.py install"

jspricke commented on cmake/templates/python_install.bat.in:
```diff
@@ -18,10 +18,5 @@ for /f "usebackq tokens=*" %%a in ('@SETUPTOOLS_INSTALL_PREFIX@') do (
-    --prefix="%_SETUPTOOLS_INSTALL_PATH%" ^
-    --install-scripts="%_SETUPTOOLS_INSTALL_PATH%\@CATKIN_GLOBAL_BIN_DESTINATION@" ^
-    --root=%_SETUPTOOLS_INSTALL_ROOT%\
+"@PYTHON_EXECUTABLE@" -m pip install --ignore-installed --no-deps --prefix "%_SETUPTOOLS_INSTALL_PATH%" ^
```
I think this needs a runtime dependency on python3-pip

v4hn replied:
```suggestion
"@PYTHON_EXECUTABLE@" -m pip install --ignore-installed --no-deps --use-pep517 --no-build-isolation --prefix "%_SETUPTOOLS_INSTALL_PATH%" ^
```

rhaschke committed:
Add pip dependency

v4hn commented:
@jspricke also pointed out that https://github.com/ros-o/catkin/pull/2 was merged before, but is not in the tree anymore for some undocumented reason. :frowning_person: 
This PR adds it back as a part, so that should be fine.

I'll build my local workspaces with the change to check whether I find any problems in the wider scope.
Thanks for working on it @rhaschke !

> Drawback: pip performs an in-place build, creating an .egg-info folder. This is deleted after installation.

Did you consider using `python -m build -w -O ${CMAKE_CURRENT_BINARY_DIR}` instead, building a wheel and storing it in build/? That should avoid the egg-info at the cost of more build overhead, but I didn't look into it further and do not know how it relates to editable installs right now.

roehling commented:
> Drawback: pip performs an in-place build, creating an .egg-info folder. This is deleted after installation.

Why remove it? Leaving it in makes e.g. entry points work (which were always an issue with catkin-based Python packages), and AFAICT it does have no adverse effects besides using up a tiny extra amount of disk space.

roehling commented on CMakeLists.txt:
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.0.2)
+cmake_minimum_required(VERSION 3.16)
```
I assume you picked 3.16 because of Ubuntu 20.04. Do we intend to support Focal past its EOL? More generally, do we have (or should we have) a policy for minimum version requirements for ROS-O?

v4hn replied:
No, there is no minimum OS version because there are no fixed target distributions, but if we can avoid breakage without consequences it makes sense to keep compatibility.
There is also no single branch for "ROS-O" and while compatibility is important, there is no problem to change branches in the individual builders for target platforms.

The question then is whether any "obese-devel"/"ros1" etc branches should move forward with breaking changes or whether we simply add more branches ("occult-devel" suggests itself ;)).

rhaschke commented:
> > Drawback: pip performs an in-place build, creating an .egg-info folder. This is deleted after installation.
> 
> Why remove it? Leaving it in makes e.g. entry points work (which were always an issue with catkin-based Python packages), and AFAICT it does have no adverse effects besides using up a tiny extra amount of disk space.

The egg-info folder is created in the source directory, where it shouldn't stay. I guess it is also installed, which is fine of course.

rhaschke commented:
> Did you consider using `python -m build -w -O ${CMAKE_CURRENT_BINARY_DIR}` instead?

Not yet. How to install from there?

v4hn commented on cmake/templates/python_install.sh.in:
```diff
@@ -22,12 +22,12 @@ echo_and_run mkdir -p "$DESTDIR@CMAKE_INSTALL_PREFIX@/@PYTHON_INSTALL_DIR@"
-    install \
-    --root="${DESTDIR-/}" \
-    @SETUPTOOLS_ARG_EXTRA@ --prefix="@CMAKE_INSTALL_PREFIX@" --install-scripts="@CMAKE_INSTALL_PREFIX@/@CATKIN_GLOBAL_BIN_DESTINATION@"
+    "@PYTHON_EXECUTABLE@" -m pip install --prefix "@CMAKE_INSTALL_PREFIX@" --ignore-installed --no-deps \
```
```suggestion
    "@PYTHON_EXECUTABLE@" -m pip install --prefix "@CMAKE_INSTALL_PREFIX@" --ignore-installed --no-deps --use-pep517 --no-build-isolation \
```
With the PR I see these deprecations over here on Debian testing:

> DEPRECATION: Building 'catkin' using the legacy setup.py bdist_wheel mechanism, which will be removed in a future version. pip 25.3 will enforce this behaviour change. A possible replacement is to use the standardized build interface by setting the `--use-pep517` option, (possibly combined with `--no-build-isolation`), or adding a `pyproject.toml` file to the source tree of 'catkin'. Discussion can be found at https://github.com/pypa/pip/issues/6334

These can be avoided by forcing the "new" pep standard, but disable build isolation because catkin_pkg (and thus other ROS components) are not available in registries but available in the workspace.

v4hn reviewed (commented):
> Not yet. How to install from there?

You should just need to specify the wheel instead of `CMAKE_CURRENT_SOURCE_DIR`:
```
pip install ${CMAKE_CURRENT_BINARY_DIR}/build_package.whl
```

Again, I did not try that approach.

> The egg-info folder is created in the source directory, where it shouldn't stay. I guess it is also installed, which is fine of course.

Did you test devel workspaces with this change?
In theory we might migrate to install --editable for them, but I'm sure there are non-trivial reasons why the approach catkin takes can be better in the ROS-O ecosystem.

My core workspace (with 417 basic packages on Debian testing - obviously not all with python setup) builds an install workspace with significantly less warnings with this patch. :heart:

rhaschke committed:
Enforcing PEP 517

... silencing the pip deprecation warning:

DEPRECATION: Building 'xxx' using the legacy setup.py bdist_wheel mechanism, which will be removed in a future version. pip 25.3 will enforce this behaviour change. A possible replacement is to use the standardized build interface by setting the `--use-pep517` option, (possibly combined with `--no-build-isolation`), or adding a `pyproject.toml` file to the source tree of 'xxx'. Discussion can be found at https://github.com/pypa/pip/issues/6334

Co-authored-by: Michael Görner <me@v4hn.de>

rhaschke committed:
Drop definition of CATKIN_BINARY_DIR

rhaschke commented:
> Did you consider using `python -m build -w -O ${CMAKE_CURRENT_BINARY_DIR}` instead?

Tried now: https://github.com/ubi-agni/catkin/compare/obese-devel...ubi-agni:catkin:python-build+pip-install
This still creates the `egg-info` folder in the source tree.

@v4hn: Could you rebuild your workspace with the latest commits?

v4hn commented on cmake/templates/python_install.sh.in:
```diff
@@ -22,12 +22,11 @@ echo_and_run mkdir -p "$DESTDIR@CMAKE_INSTALL_PREFIX@/@PYTHON_INSTALL_DIR@"
-    --root="${DESTDIR-/}" \
-    @SETUPTOOLS_ARG_EXTRA@ --prefix="@CMAKE_INSTALL_PREFIX@" --install-scripts="@CMAKE_INSTALL_PREFIX@/@CATKIN_GLOBAL_BIN_DESTINATION@"
+    "@PYTHON_EXECUTABLE@" -m pip install --prefix "@CMAKE_INSTALL_PREFIX@" --ignore-installed --no-deps --use-pep517 --no-build-isolation \
+    "@CMAKE_CURRENT_SOURCE_DIR@"
```
This creates empty `build/` folders in each project dir.
Here is a workaround to have pip forward the `build` command to setup.py with an option to use a different build directory in the workspace build folder
(also needed in the bat if we go for that road)

```suggestion
    "@PYTHON_EXECUTABLE@" -m pip install \
      --prefix "@CMAKE_INSTALL_PREFIX@" \
      --ignore-installed \
      --no-deps \
      --use-pep517 \
      --no-build-isolation \
      --config-settings=--global-option=build \
      --config-settings=--global-option=--build-base=@CMAKE_BINARY_DIR@/setuptools_build \
      "@CMAKE_CURRENT_SOURCE_DIR@"
```

Also, I tested devel workspaces for a bit and the previous logic there remains working.

v4hn reviewed (commented):
> Why remove [the egg-info in src]? Leaving it in makes e.g. entry points work (which were always an issue with catkin-based Python packages), and AFAICT it does have no adverse effects besides using up a tiny extra amount of disk space.

@roehling The whole install logic only touches install workspaces, which need to be standalone and not devel workspaces, where it might make sense to keep files in `src/` if it can help python.
It might make sense to use editable installs with devel workspaces and leave the `egg-info` there, but catkin uses a different awkward mechanism by overlaying all python modules/scripts with wrapper files in devel.

> I guess it is also installed, which is fine of course.

@rhaschke I guess you mean the installed `*.dist-info` in install/?

Crucially, I went on to build my MoveIt workspaces now and hit `py_binding_tools` which still fails due to a bug in pybind11 [that you reported upstream last year with still no response](https://github.com/ros/catkin/pull/1195#pullrequestreview-2514069334).
You also documented there that you removed the corresponding ros-o commit at that point.
That still seems like a major deal-breaker. Do we have any reasonable options to address this in the moveit repositories?

v4hn commented:
Status Update.

I started working on a buildfarm-integration of this effort and several small things did not perfectly work out yet. There are several repositories involved.

On my side [this branch](https://github.com/v4hn/ros-o-builder/tree/build-for-bookworm-modernize-python) currently describes the automated build attempt for Debian bookworm (I know, trixi is next), listing the repositories and modified branches in the latest commits.
